blog/modules/ROOT/pages/class-as-a-function.adoc

```diff
@@ -0,0 +1,78 @@
+= Class-as-a-Function Pattern
+
+:use-link-attrs:
+
+// tag::byline[]
+++++
+<div class="blog-byline">
+++++
+by link:https://github.com/HT154[Josh Basch] on August 6th, 2024
+++++
+</div>
+++++
+// end::byline[]
+
+// tag::excerpt[]
+In many languages, function and method parameters may be assigned default values allowing the argument to be omitted in calls.
+Many languages also offer named parameters, which aid in API self-documentation.
+Pkl xref:main:language-reference:index.adoc#methods[methods] do not provide either feature, but one way to achieve similar results is the "class-as-a-function" (CaaF) pattern.
```

Comment:
Non-blocking comment:

Something else that this pattern lets you do is to build abstraction layers. For example, you can build a higher level api for a Kubernetes deployment where you provide knobs for the image to deploy, the healthcheck endpoint, and your `result` is a `k8s.api.apps.v1.Deployment` with all the sidecars/liveness probes set up for you.

Sometimes we call this "embedding" a class, because your class is a higher level version of the embedded class definition.

I think it'd be really useful to have some sort of blog post that also details this, but this might make your blog post a lot longer. We don't need to block on this if it's a bigger bite than you'd like to handle.

Reply:
I actually had an example half-drafted and waffled on including this. I've fleshed it out and added it!